intake_netflow/v9.py

```diff
@@ -0,0 +1,520 @@
+    def decode(source):
+        template_id, nfields = read_and_unpack(source, s_type_length)
+
+        template = TemplateRecord(template_id)
+        for _ in range(nfields):
+            template.fields.append(TemplateField.decode(source))
+
+        return template
+
+    def encode(self):
+        raw = s_type_length.pack(self.id, len(self.fields))
+        for field in self.fields:
+            raw += field.encode()
+        return raw
+
+
+class TemplateFlowSet(object):
+    def __init__(self, templates=None):
+        self.id = 0
+        self.templates = {template.id: template for template in templates} if templates else {}
+
+    def __eq__(self, other):
+        return self.id == other.id and self.templates == other.templates
+
+    def __len__(self):
+        nbytes = s_type_length.size
+        for template in self.templates.values():
+            nbytes += len(template)
+        return nbytes
+
+    def __getitem__(self, key):
+        return self.templates[key]
+
+    def __iter__(self):
+        return iter(self.templates)
+
+    @staticmethod
+    def decode(source):
+        fs = TemplateFlowSet()
+        _, length = read_and_unpack(source, s_type_length)
+        offset = s_type_length.size
+
+        while offset < length:
+            template = TemplateRecord.decode(source)
+            fs.templates[template.id] = template
+            offset += len(template)
+
+        return fs
+
+    def encode(self):
+        raw = s_type_length.pack(self.id, len(self))
+        for template in self.templates.values():
+            raw += template.encode()
+        return raw
+
+
+class DataFlowSet(object):
+    def __init__(self, id, payload, templates):
+        self.template = templates[id]
+        self.records = []
+        self.record_length = len(self.template) - s_type_length.size
+
+        if isinstance(payload, bytes):
+            source = byte_stream(payload)
+            remaining = len(payload)
+            while remaining >= self.record_length:
+                self.records.append([read_and_unpack(source, field.struct)[0] for field in self.template])
+                remaining -= self.record_length
+        elif isinstance(payload, list):
+            self.records = payload
+
+    def __len__(self):
+        return s_type_length.size + len(self.records) * self.record_length
+
+    def __iter__(self):
+        return iter(self.records)
+
+    @staticmethod
+    def decode(source):
+        id, length = read_and_unpack(source, s_type_length)
+        payload = source.read(length - s_type_length.size)
+        return functools.partial(DataFlowSet, id, payload)
+
+    def encode(self):
+        raw = s_type_length.pack(self.template.id, len(self))
+        for record in self.records:
+            for field, value in zip(self.template.fields, record):
+                raw += field.struct.pack(value)
+        return raw
+
+
+def decode_flowset(source):
+    raw = source.peek(s_flowset.size)[:s_flowset.size]
```

Comment:
Why would the return from `peek` ever be more than the requested number of bytes? In any case, I don't think all file-system backends support it, so would be better to have 
```python
loc = source.tell()
raw = source.read(s_flowset.size)
source.seek(loc)
```

Reply:
`io.BufferedReader.peek` allows less or more than the requested bytes to be returned.

Since we need to support file-like objects instead, I replaced the `peek` method with a local one using the `tell`/`seek` combo.